yearn/treasury/treasury.py

```diff
@@ -0,0 +1,319 @@
+        cdp_manager = Contract('0x5ef30b9986345249bc32d8928B7ee64DE9435E39')
+        # ychad = Contract('ychad.eth')
+        ychad = Contract('0xfeb4acf3df3cdea7399794d0869ef76a6efaff52')
+        vat = Contract('0x35D1b3F3D7966A1DFe207aa4514C12a259A0492B')
+        proxy = proxy_registry.proxies(ychad)
+        cdp = cdp_manager.first(proxy)
+        urn = cdp_manager.urns(cdp)
+        ilk = encode_single('bytes32', b'YFI-A')
+        art = vat.urns(ilk, urn, block_identifier=block).dict()["art"]
+        rate = vat.ilks(ilk, block_identifier=block).dict()["rate"]
+        debt = art * rate / 1e45
+        dai = '0x6B175474E89094C44Da98b954EedeAC495271d0F'
+        debt = {dai: {'balance': debt, 'usd value': debt}}
+        return debt
+
+    def unit_debt(self, block=None) -> dict:
+        if block and block < 11315910:
+            return
+        # ychad = Contract('ychad.eth')
+        ychad = Contract('0xfeb4acf3df3cdea7399794d0869ef76a6efaff52')
+        unitVault = Contract("0xb1cff81b9305166ff1efc49a129ad2afcd7bcf19")
+        yfi = "0x0bc529c00C6401aEF6D220BE8C6Ea1667F6Ad93e"
+        usdp = '0x1456688345527bE1f37E9e627DA0837D6f08C925'
+        debt = unitVault.getTotalDebt(yfi, ychad, block_identifier=block) / 10 ** 18
+        debt = {usdp: {'balance': debt, 'usd value': debt}}
+        return debt
+
+    # helper functions
+
+    def load_transfers(self):
+        if not self._thread._started.is_set():
+            self._thread.start()
+        self._done.wait()
+
+    def watch_transfers(self):
+        start = time.time()
+        logger.info(
+            'pulling treasury transfer events, please wait patiently this takes a while...'
+        )
+        # Treasury didn't exist prior to block 10502337
+        self.log_filter_in = web3.eth.filter({"fromBlock": 10502337, "topics": self._topics_in})
+        self.log_filter_out = web3.eth.filter({"fromBlock": 10502337, "topics": self._topics_out})
+        for block in chain.new_blocks(height_buffer=12):
+            logs = self.log_filter_in.get_new_entries()
+            self.process_transfers(logs)
+            logs = self.log_filter_out.get_new_entries()
+            self.process_transfers(logs)
+            if not self._done.is_set():
+                self._done.set()
+                logger.info(
+                    "loaded treasury transfer events in %.3fs", time.time() - start
+                )
+            if not self._watch_events_forever:
+                break
+            time.sleep(30)
+
+    def process_transfers(self, logs):
+        for log in logs:
+            try:
+                event = decode_logs(
+                    [log]
+                )  # NOTE: We have to decode logs here because NFTs prevent us from batch decoding logs
+                self._transfers.append(event)
+            except:
+                if log.address in [
+                    # These are NFTs # TODO 
```

Comment:
maybe there could be some heuristic to detect nfts?

Reply:
Agreed, I had wanted to see a few more instances of NFTs that cause problems before I try to extrapolate from the way these particular logs are formed